README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# metawho

[![CRAN
status](https://www.r-pkg.org/badges/version/metawho)](https://cran.r-project.org/package=metawho)
[![](http://cranlogs.r-pkg.org/badges/grand-total/metawho?color=blue)](https://cran.r-project.org/package=metawho)
[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![AppVeyor build
status](https://ci.appveyor.com/api/projects/status/github/ShixiangWang/metawho?branch=master&svg=true)](https://ci.appveyor.com/project/ShixiangWang/metawho)
[![Travis build
status](https://travis-ci.org/ShixiangWang/metawho.svg?branch=master)](https://travis-ci.org/ShixiangWang/metawho)
[![Coverage
status](https://codecov.io/gh/ShixiangWang/metawho/branch/master/graph/badge.svg)](https://codecov.io/github/ShixiangWang/metawho?branch=master)

The goal of **metawho** is to provide simple R implementation of
“Meta-analytical method to Identify Who Benefits Most from Treatments”
(called ‘deft’ approach, see [reference \#2](#references)).

**metawho** is powered by R package **metafor** and does not support
dataset contains individuals for now. Please use stata package
**ipdmetan** if you are more familar with stata code.

## Installation

You can install the development version of **metawho** from GitHub with:

``` r
remotes::install_github("ShixiangWang/metawho")
```

Or install stable release from CRAN.

## References

If you use this package in academic field, please cite the following
papers:

  - Wang, Shixiang, et al. “The predictive power of tumor mutational
    burden in lung cancer immunotherapy response is influenced by
    patients’ sex.” International journal of cancer (2019).
  - Fisher, David J., et al. “Meta-analytical methods to identify who
    benefits most from treatments: daft, deluded, or deft approach?.”
    bmj 356 (2017): j573.


NEWS.md
# metawho 0.2.0

* Fixed some small bugs.
* Added `deft_show()` to show deft results.

# metawho 0.1.0

* Submitted to CRAN
* Updated setting in `rma()` function & fixed overall result not show in figure.
* Added function to reproduce results in wang et al (2019).
* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(metawho)

### specify hazard ratios (hr)
hr <- c(0.30, 0.11, 1.25, 0.63, 0.90, 0.28)
### specify lower bound for hr confidence intervals
ci.lb <- c(0.09, 0.02, 0.82, 0.42, 0.41, 0.12)
### specify upper bound for hr confidence intervals
ci.ub <- c(1.00, 0.56, 1.90, 0.95, 1.99, 0.67)
### specify sample number
ni <- c(16L, 18L, 118L, 122L, 37L, 38L)
### trials
trial <- c(
  "Rizvi 2015", "Rizvi 2015",
  "Rizvi 2018", "Rizvi 2018",
  "Hellmann 2018", "Hellmann 2018"
)
### subgroups
subgroup <- rep(c("Male", "Female"), 3)

entry <- paste(trial, subgroup, sep = "-")
### combine as data.frame

wang2019 <-
  data.frame(
    entry = entry,
    trial = trial,
    subgroup = subgroup,
    hr = hr,
    ci.lb = ci.lb,
    ci.ub = ci.ub,
    ni = ni,
    stringsAsFactors = FALSE
  )

deft_prepare(wang2019)

testthat::expect_true(is.data.frame(wang2019))

data("wang2019")
res <- deft_do(wang2019, group_level = c("Male", "Female"))

testthat::expect_s3_class(res, "deft")

p1 <- deft_show(res, "all")
p2 <- deft_show(res, "subgroup")

testthat::expect_s3_class(p1, "ggplot")
testthat::expect_s3_class(p2, "ggplot")


R/deft_do.R
#' Implement deft method
#'
#' 'deft' method is a meta-analytical approach to pool conclusion from multiple
#' studies. More details please see references.
#'
#' About model fit, please see [metafor::rma()].
#'
#' @param prepare a result `data.frame` from [deft_prepare] function or
#' a `data.frame` contains at least 'trial', 'subgroup', 'yi' and 'sei' these
#' four columns.
#' @param group_level level of subgroup, should be a character vector with
#' length 2 and the reference should put in the first. For example, if you
#' have 'Male' and 'Female' groups and want compare 'Female' with 'Male', then
#' should set `c('Male', 'Female')`.
#' @inheritParams metafor::rma
#' @author Shixiang Wang <w_shixiang@163.com>
#' @references Fisher, David J., et al. "Meta-analytical methods to identify who
#' benefits most from treatments: daft, deluded, or deft approach?." bmj 356 (2017): j573.
#'
#' Wang, Shixiang, et al. "The predictive power of tumor mutational burden
#' in lung cancer immunotherapy response is influenced by patients' sex."
#' International journal of cancer (2019).
#' @return a `list` which class is 'deft'.
#' @import metafor
#' @export
#'
#' @examples
#' data("wang2019")
#' deft_do(wang2019, group_level = c("Male", "Female"))
deft_do <- function(prepare, group_level, method = "FE") {
  stopifnot(is.data.frame(prepare), is.character(group_level))
  if (!all(c("trial", "subgroup", "yi", "sei") %in% colnames(prepare))) {
    msg <- paste("Four columns called trial, subgroup, hr, yi and sei must in input data!",
      "==============",
      "Example data please run data('wang2019')",
      sep = "\n"
    )
    stop(msg)
  }
  if (!"entry" %in% colnames(prepare)) {
    prepare <- prepare %>%
      dplyr::mutate(entry = paste(trial, subgroup, sep = "-"))
  }

  prepare[["subgroup"]] <- factor(prepare[["subgroup"]], levels = group_level)
  before <- list()
  before[["data"]] <- prepare
  before[["model"]] <- rma(yi = yi, sei = sei, ni = ni, data = prepare, method = method)
  # Step 1:
  # divide trial into groups and build models
  f <- factor(prepare$trial, levels = unique(prepare$trial))
  model_list <- prepare %>%
    split(f) %>%
    purrr::map(~ rma(yi ~ subgroup, sei = sei, ni = ni, data = ., method = "FE")) # Only can use FE here?

  # Step 2:
  # pool results from Step 1
  after <- list()
  after[["data"]] <- purrr::map2_df(model_list, names(model_list), function(x, y) {
    s <- summary(x)
    data.frame(
      trial = y,
      hr = exp(s[["b"]][2]),
      ci.lb = exp(s[["ci.lb"]][2]),
      ci.ub = exp(s[["ci.ub"]][2]),
      ni = sum(s[["ni"]], na.rm = TRUE),
      stringsAsFactors = FALSE
    )
  })
  after[["data"]] <- deft_prepare(after[["data"]])
  after[["model"]] <- rma(
    yi = yi, sei = sei, ni = ni,
    data = after[["data"]], method = method
  )

  res <- list()
  res[["all"]] <- before
  res[["subgroup"]] <- after
  class(res) <- "deft"
  res
}




#-----------------
utils::globalVariables(
  c(".", "ci.lb", "ci.ub", "conf_q", "hr", "yi", "sei", "subgroup", "trial", "ni")
)


R/deft_show.R
#' Show deft result
#'
#' @param deft result from [deft_do].
#' @param element 'all' or 'subgroup'.
#' @param study_labels labels for studies.
#' @param headings a list for controlling plot headings.
#' @param ... other arguments except 'panels', 'trans', 'study_labels',
#' and 'show_stats' passed to [forestmodel::forest_rma()].
#' @inheritParams forestmodel::forest_rma
#'
#' @return a `ggplot` object
#' @export
#'
#' @examples
#' data("wang2019")
#' res <- deft_do(wang2019, group_level = c("Male", "Female"))
#'
#' p1 <- deft_show(res, "all")
#' p1
#'
#' p2 <- deft_show(res, "subgroup")
#' p2
deft_show <- function(deft, element, study_labels = NULL,
                      headings = list(
                        study = ifelse(element == "all", "Study-subgroup", "Study"),
                        n = "N", measure = NULL, ci = "HR (95% CI)"
                      ),
                      trans = base::exp,
                      show_model = ifelse(element == "all", FALSE, TRUE),
                      show_stats = list(
                        `I^2` = rlang::quo(sprintf("%0.1f%%", I2)),
                        p = rlang::quo(format.pval(QEp,
                          digits = 2
                        ))
                      ),
                      ...) {
  stopifnot(inherits(deft, "deft"))
  df <- deft[[element]]
  if (is.null(study_labels)) {
    if (element == "all") {
      study_labels <- df$data$entry
    } else {
      study_labels <- df$data$trial
    }
  }

  p <- df$model %>%
    forestmodel::forest_rma(
      panels = deft_panel(.,
        headings = headings
      ),
      trans = trans,
      study_labels = study_labels,
      show_stats = show_stats,
      show_model = show_model,
      ...
    )
  p
}

#' @import magrittr forestmodel
deft_panel <- function(model = NULL, factor_separate_line = FALSE,
                       headings = list(study = "Study", n = "N", measure = "HR", ci = NULL),
                       trans_char = "I") {
  if (inherits(model, "rma")) {
    if (trans_char == "I") {
      trans_char <- "FALSE"
    }

    panels <- list(
      forest_panel(width = 0.03),
      forest_panel(
        width = 0.01, display = study, fontface = "bold", heading = headings$study,
        width_group = 1
      ),
      forest_panel(
        width = 0.18, display = stat, parse = TRUE,
        width_group = 1
      ),
      forest_panel(width = 0.05, display = n, hjust = 1, heading = headings$n),
      forest_panel(width = 0.03, item = "vline", hjust = 0.5),
      forest_panel(
        width = 0.55, item = "forest", hjust = 0.5, heading = headings$measure,
        linetype = "dashed", line_x = 0
      ),
      forest_panel(width = 0.03, item = "vline", hjust = 0.5),
      forest_panel(
        width = 0.12,
        display = sprintf("%0.2f (%0.2f, %0.2f)", trans(estimate), trans(conf.low), trans(conf.high)),
        heading = headings$ci,
        display_na = NA
      ),
      forest_panel(width = 0.03)
    )
  } else {
    stop("This function only support rma object.")
  }
  panels
}

utils::globalVariables(
  c("I2", "QEp", "conf.high", "conf.low", "estimate", "n", "stat", "study", "trans")
)


inst/develop.R
library(metawho)
#> Loading required package: metafor
#> Loading required package: Matrix
#> Loading 'metafor' package (version 2.0-0). For an overview
#> and introduction to the package please type: help(metafor).

### specify hazard ratios (hr)
hr    <- c(0.30, 0.11, 1.25, 0.63, 0.90, 0.28)
### specify lower bound for hr confidence intervals
ci.lb <- c(0.09, 0.02, 0.82, 0.42, 0.41, 0.12)
### specify upper bound for hr confidence intervals
ci.ub <- c(1.00, 0.56, 1.90, 0.95, 1.99, 0.67)
ni <- c(16L, 18L, 118L, 122L, 37L, 38L)
### trials
trial <- c("Rizvi 2015", "Rizvi 2015",
           "Rizvi 2018", "Rizvi 2018",
           "Hellmann 2018", "Hellmann 2018")
### subgroups
subgroup = rep(c("Male", "Female"), 3)

entry <- paste(trial, subgroup, sep = "-")
### combine as data.frame

wang2019 =
    data.frame(
        entry = entry,
        trial = trial,
        subgroup = subgroup,
        hr = hr,
        ci.lb = ci.lb,
        ci.ub = ci.ub,
        ni = ni,
        stringsAsFactors = FALSE
    )

library(metawho)

pre = deft_prepare(wang2019)
(res = deft_do(pre, group_level = c("Male", "Female"), method = "DL"))
library(forestmodel)
forestmodel::forest_rma(list(res$all$model, res$subgroup$model), trans=exp,
                        panels = default_forest_panels(res$all$model,
                        factor_separate_line = factor_separate_line,
                        measure = "HR"),
                        model_label = c("What", "here")
                        )
forestmodel::forest_rma(res$subgroup$model, trans=exp)

deft_show(res, "all")
deft_show(res, "subgroup")


R/deft_prepare.R
#' @title Prepare log transformation data for effect size estimation
#' according to confidence level and distribution
#'
#' @description A variety of different outcome measures which
#' used in meta-analysis as input are in the form of log, such as hazard ratio (HR).
#' This function is used to do log transformation to calculate effect size and
#' standard error. Then the result can be easier used for model fit.
#'
#' @param data  a `data.frame` contains at least columns 'trial', 'hr', 'ci.lb',
#' 'ci.ub' and 'ni'.
#' @param conf_level a number specify confidence level, default is 0.05.
#' @author Shixiang Wang <w_shixiang@163.com>
#' @return a `data.frame`
#' @references Wang, Shixiang, et al. "The predictive power of tumor mutational burden
#' in lung cancer immunotherapy response is influenced by patients' sex."
#' International journal of cancer (2019).
#' @examples
#' ### specify hazard ratios (hr)
#' hr <- c(0.30, 0.11, 1.25, 0.63, 0.90, 0.28)
#' ### specify lower bound for hr confidence intervals
#' ci.lb <- c(0.09, 0.02, 0.82, 0.42, 0.41, 0.12)
#' ### specify upper bound for hr confidence intervals
#' ci.ub <- c(1.00, 0.56, 1.90, 0.95, 1.99, 0.67)
#' ### specify sample number
#' ni <- c(16L, 18L, 118L, 122L, 37L, 38L)
#' ### trials
#' trial <- c(
#'   "Rizvi 2015", "Rizvi 2015",
#'   "Rizvi 2018", "Rizvi 2018",
#'   "Hellmann 2018", "Hellmann 2018"
#' )
#' ### subgroups
#' subgroup <- rep(c("Male", "Female"), 3)
#'
#' entry <- paste(trial, subgroup, sep = "-")
#' ### combine as data.frame
#'
#' wang2019 <-
#'   data.frame(
#'     entry = entry,
#'     trial = trial,
#'     subgroup = subgroup,
#'     hr = hr,
#'     ci.lb = ci.lb,
#'     ci.ub = ci.ub,
#'     ni = ni,
#'     stringsAsFactors = FALSE
#'   )
#'
#' deft_prepare(wang2019)
#' @import stats
#' @export
deft_prepare <- function(data, conf_level = 0.05) {
  stopifnot(is.data.frame(data))
  if (!all(c("hr", "ci.lb", "ci.ub", "trial", "ni") %in% colnames(data))) {
    msg <- paste("Five columns called trial, hr, ci.lb, ci.ub and ni must in input data:",
      "    -- trial    : name or id for trial",
      "    -- hr       : hazard ratio",
      "    -- ci.lb    : lower bound of confidence interval",
      "    -- ci.ub    : upper bound of confidence interval",
      "    -- ni       : sample number",
      sep = "\n"
    )
    stop(msg)
  }


  dat <- dplyr::mutate(
    data,
    conf_q = qnorm(1 - conf_level / 2),
    yi = log(hr),
    sei = (log(ci.ub) - log(ci.lb)) / (2 * conf_q),
    ni = .data$ni
  )

  return(dat)
}


data-raw/example_data.R
#' @source Wang, Shixiang, Jing Zhang, Zaoke He, Kai Wu, and Xue‐Song Liu.
#' “The Predictive Power of Tumor Mutational Burden in Lung Cancer Immunotherapy Response Is
#' Influenced by Patients’ Sex.” International Journal of Cancer, April 29, 2019,
#' ijc.32327. https://doi.org/10.1002/ijc.32327.

library(metawho)

### specify hazard ratios (hr)
hr <- c(0.30, 0.11, 1.25, 0.63, 0.90, 0.28)
### specify lower bound for hr confidence intervals
ci.lb <- c(0.09, 0.02, 0.82, 0.42, 0.41, 0.12)
### specify upper bound for hr confidence intervals
ci.ub <- c(1.00, 0.56, 1.90, 0.95, 1.99, 0.67)
### specify sample number
ni <- c(16L, 18L, 118L, 122L, 37L, 38L)
### trials
trial <- c(
  "Rizvi 2015", "Rizvi 2015",
  "Rizvi 2018", "Rizvi 2018",
  "Hellmann 2018", "Hellmann 2018"
)
### subgroups
subgroup <- rep(c("Male", "Female"), 3)

entry <- paste(trial, subgroup, sep = "-")
### combine as data.frame

wang2019 <-
  data.frame(
    entry = entry,
    trial = trial,
    subgroup = subgroup,
    hr = hr,
    ci.lb = ci.lb,
    ci.ub = ci.ub,
    ni = ni,
    stringsAsFactors = FALSE
  )

wang2019 <- deft_prepare(wang2019)
usethis::use_data(wang2019, overwrite = TRUE)


R/main.R
#'@title Prepare log transformation data for effect size estimation
#'according to confidence level and distribution
#'
#'@description A variety of different outcome measures which
#'used in meta-analysis as input are in the form of log, such as hazard ratio (HR).
#'This function is used to do log transformation to calculate effect size and
#'standard error. Then the result can be easier used for model fit.
#'
#'@param data  a `data.frame` contains at least columns 'trial', 'hr', 'ci.lb', 'ci.ub'.
#'@param distribution a character specify distribution.
#''N' for normal, 't' for student distribution.
#'Default is N for normal distribution.
#'@param conf_level a number specify confidence level, default is 0.05.
#'@param df a number specify degree of freedom for t distribution
#'@param var default is FALSE. If TRUE, the sampling variance will be computed.
#'@author Shixiang Wang <w_shixiang@163.com>
#'@return a `data.frame`
#'@references Wang, Shixiang, et al. "The predictive power of tumor mutational burden
#' in lung cancer immunotherapy response is influenced by patients' sex."
#' International journal of cancer (2019).
#'@examples
#'### specify hazard ratios (hr)
#'hr    <- c(0.30, 0.11, 1.25, 0.63, 0.90, 0.28)
#'### specify lower bound for hr confidence intervals
#'ci.lb <- c(0.09, 0.02, 0.82, 0.42, 0.41, 0.12)
#'### specify upper bound for hr confidence intervals
#'ci.ub <- c(1.00, 0.56, 1.90, 0.95, 1.99, 0.67)
#'### trials
#'trial <- c("Rizvi 2015", "Rizvi 2015",
#'           "Rizvi 2018", "Rizvi 2018",
#'           "Hellmann 2018", "Hellmann 2018")
#'### subgroups
#'subgroup = rep(c("Male", "Female"), 3)
#'
#'entry <- paste(trial, subgroup, sep = "-")
#'### combine as data.frame
#'
#'wang2019 =
#'    data.frame(
#'         entry = entry,
#'         trial = trial,
#'         subgroup = subgroup,
#'         hr = hr,
#'         ci.lb = ci.lb,
#'         ci.ub = ci.ub,
#'         stringsAsFactors = FALSE
#'        )
#'
#'deft_prepare(wang2019)

#'@import stats
#'@export
deft_prepare = function(data, distribution=c("N", "t"),
                  conf_level=0.05,
                  df=Inf,
                  var=FALSE){

    stopifnot(is.data.frame(data))
    distribution = match.arg(distribution)
    if (!all(c("hr", "ci.lb", "ci.ub", "trial") %in% colnames(data))) {
        msg = paste("Four columns called trial, hr, ci.lb and ci.ub must in input data:",
                    "    -- trial    : name or id for trial",
                    "    -- hr       : hazard ratio",
                    "    -- ci.lb    : lower bound of confidence interval",
                    "    -- ci.ub    : upper bound of confidence interval",
                    sep = "\n")
        stop(msg)
    }

    if(distribution == "N"){
        dat = dplyr::mutate(
            data,
            conf_q = qnorm(1 - conf_level/2),
            yi  = log(hr),
            sei  = (log(ci.ub) - log(ci.lb)) / (2*conf_q)
        )


    }else if (distribution == "t"){
        dat = dplyr::mutate(
            data,
            conf_q = qt(1 - conf_level/2, df = df),
            yi  = log(hr),
            sei  = (log(ci.ub) - log(ci.lb)) / (2*conf_q)
        )

    }

    if (var) {
        dat = dplyr::mutate(
            vi = sei^2
        )
    }

    return(dat)
}


#' Implement deft method
#'
#' 'deft' method is a meta-analytical approach to pool conclusion from multiple
#' studies. More details please see references.
#'
#' About model fit, please see [metafor::rma()].
#'
#' @param prepare a result `data.frame` from [deft_prepare] function or
#' a `data.frame` contains at least 'trial', 'subgroup', 'yi' and 'sei' these
#' four columns.
#' @param group_level level of subgroup, should be a character vector with
#' length 2 and the reference should put in the first. For example, if you
#' have 'Male' and 'Female' groups and want compare 'Female' with 'Male', then
#' should set `c('Male', 'Female')`.
#' @inheritParams metafor::rma
#' @author Shixiang Wang <w_shixiang@163.com>
#' @references Fisher, David J., et al. "Meta-analytical methods to identify who
#' benefits most from treatments: daft, deluded, or deft approach?." bmj 356 (2017): j573.
#'
#' Wang, Shixiang, et al. "The predictive power of tumor mutational burden
#' in lung cancer immunotherapy response is influenced by patients' sex."
#' International journal of cancer (2019).
#' @return a `list` which class is 'deft'.
#' @import metafor
#' @export
#'
#' @examples
#' data('wang2019')
#' deft_do(wang2019, group_level = c("Male", "Female"))
deft_do = function(prepare, group_level, method = "FE") {

    stopifnot(is.data.frame(prepare), is.character(group_level))
    if (!all(c("trial", "subgroup", "yi", "sei") %in% colnames(prepare))) {
        msg = paste("Four columns called trial, subgroup, hr, yi and sei must in input data!",
                    "==============",
                    "Example data please run data('wang2019')",
                    sep = "\n")
        stop(msg)
    }
    if (! 'entry' %in% colnames(prepare)) {
        prepare = dplyr::mutate(
            entry = paste(trial, subgroup, sep = '-')
        )
    }

    prepare[["subgroup"]] = factor(prepare[["subgroup"]], levels = group_level)
    before = list()
    before[["data"]] = prepare
    before[["model"]] = rma(yi = yi, sei = sei, data = prepare, method = method)
    # Step 1:
    # divide trial into groups and build models
    model_list = prepare %>% split(.$trial) %>%
        purrr::map(~rma(yi ~ subgroup, sei = sei, data = ., method = "FE")) # Only can use FE here?

    # Step 2:
    # pool results from Step 1
    after = list()
    after[["data"]] = purrr::map2_df(model_list, names(model_list), function(x, y){
        s = summary(x)
        data.frame(
            trial = y,
            hr = exp(s[['b']][2]),
            ci.lb = exp(s[['ci.lb']][2]),
            ci.ub = exp(s[['ci.ub']][2]),
            stringsAsFactors = FALSE
        )
    })
    after[["data"]] = deft_prepare(after[["data"]])
    after[["model"]] = rma(yi = yi, sei = sei,
                           data = after[["data"]], method = method)

    res = list()
    res[["all"]] = before
    res[["subgroup"]] = after
    class(res) = 'deft'
    res
}




#-----------------
utils::globalVariables(
    c(".", "ci.lb", "ci.ub", "conf_q", "hr", "yi", "sei", "subgroup", "trial")
)
